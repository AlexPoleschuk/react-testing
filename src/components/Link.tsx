import {
    type FC,
    type ReactNode,
    useState,
} from 'react';

const STATUS = {
    HOVERED: 'hovered',
    NORMAL: 'normal',
};

interface LinkPropsType {
    children: ReactNode;
    url?: string;
}

const Link: FC<LinkPropsType> = ({
    children,
    url = '#',
}) => {
    const [status, setStatus] = useState(STATUS.NORMAL);

    const onMouseEnter = () => {
        setStatus(STATUS.HOVERED);
    };

    const onMouseLeave = () => {
        setStatus(STATUS.NORMAL);
    };

    return (
        <a
            className={status}
            role={'test_link'}
            href={url}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
        >
            {children}
        </a>
    );
};

export default Link;
