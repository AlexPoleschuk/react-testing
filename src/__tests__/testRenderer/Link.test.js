import { create, act } from 'react-test-renderer';

import Link from '../../components/Link';

describe('Link', () => {
    test('1) is component rendered', () => {
        const component = create(
            <Link url={"test_url"}>
                link content
            </Link>
        );

        const testInstance = component.root;

        expect(testInstance.findByType(Link).props.url).toBe("test_url");
        expect(testInstance.findByType(Link).props.children).toBe("link content");
    });

    test('2) is component class changed after mouse events', () => {
        const component = create(
            <Link url={"test_url"}>
                link content
            </Link>
        );

        const testInstance = component.root;
        let tree = component.toJSON();

        expect(testInstance.findByType('a').props.className).toBe("normal");

        act(() => {
            tree.props.onMouseEnter();
        });

        expect(testInstance.findByType('a').props.className).toBe("hovered");

        act(() => {
            tree.props.onMouseLeave();
        });

        expect(testInstance.findByType('a').props.className).toBe("normal");
    });

    test('3) making snapshots with mouse events', () => {
        const component = create(
            <Link url={"test_url"}>
                link content
            </Link>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        act(() => {
            tree.props.onMouseEnter();
        });

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();


        act(() => {
            tree.props.onMouseLeave();
        });

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
