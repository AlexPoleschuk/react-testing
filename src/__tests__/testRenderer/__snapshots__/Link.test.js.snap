// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link 3) making snapshots with mouse events 1`] = `
<a
  className="normal"
  href="test_url"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="test_link"
>
  link content
</a>
`;

exports[`Link 3) making snapshots with mouse events 2`] = `
<a
  className="hovered"
  href="test_url"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="test_link"
>
  link content
</a>
`;

exports[`Link 3) making snapshots with mouse events 3`] = `
<a
  className="normal"
  href="test_url"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="test_link"
>
  link content
</a>
`;
